volatility3/framework/plugins/linux/kmsg.py

```diff
@@ -186,69 +185,132 @@ def get_facility_text(cls, facility: int) -> str:
             return str(facility)
 
 
-class KmsgLegacy(ABCKmsg):
-    """Linux kernels prior to v5.10, the ringbuffer is initially kept in
-    __log_buf, and log_buf is a pointer to the former. __log_buf is declared as
-    a char array but it actually contains an array of printk_log structs.
-    The length of this array is defined in the kernel KConfig configuration via
-    the CONFIG_LOG_BUF_SHIFT value as a power of 2.
-    This can also be modified by the log_buf_len kernel boot parameter.
-    In SMP systems with more than 64 CPUs this ringbuffer size is dynamically
-    allocated according the number of CPUs based on the value of
-    CONFIG_LOG_CPU_MAX_BUF_SHIFT, and the log_buf pointer is updated
-    consequently to the new buffer.
-    In that case, the original static buffer in __log_buf is unused.
+class Kmsg_pre_3_5(ABCKmsg):
+    """The kernel ring buffer (log_buf) is a char array that sequentially stores
+    log lines, each separated by newline (LF) characters. i.e:
+        <6>[ 9565.250411] line1!\n<6>[ 9565.250412] line2\n...
     """
 
     @classmethod
     def symtab_checks(cls, vmlinux) -> bool:
-        return vmlinux.has_type("printk_log")
+        return (
+            vmlinux.has_symbol("log_end")
+            and not vmlinux.has_symbol("log_first_idx")
+            and not (
+                vmlinux.has_type("log")
+                and vmlinux.get_type("log").has_member("ts_nsec")
+            )
+        )
+
+    def run(self) -> Iterator[Tuple[str, str, str, str, str]]:
```

Comment:
The rest is so well docstringed, but it's not immediately obvious what the tuple contains?  Any chance of documenting exactly what's in a `klog`, given that's all the `run` and `run_all` functions of ABCKmsg say they return?

Reply:
Thanks @ikelos. Please check if you are happy with the latest changes